repose-aggregator/tests/functional-tests/src/integrationTest/groovy/features/filters/samlpolicy/SamlFlow20Test.groovy

```diff
@@ -0,0 +1,421 @@
+
+        and: "the origin service received the request with the correct header values"
+        mc.handlings[0]
+        mc.handlings[0].request.headers.getFirstValue(CONTENT_TYPE) == CONTENT_TYPE_XML
+        mc.handlings[0].request.headers.getFirstValue(IDENTITY_API_VERSION) == "2.0"
+        fakeIdentityV2Service.getGenerateTokenFromSamlResponseCount() == 1
+
+        when: "the saml:response received by the origin service is unmarshalled"
+        Response response = samlUtilities.unmarshallResponse(mc.handlings[0].request.body as String)
+
+        then: "there should be four assertions"
+        response.assertions.size() == 4
+
+        and: "the first assertion and the saml:response should indicate that Repose is the Issuer"
+        response.assertions[0].issuer.value == SAML_REPOSE_ISSUER
+        response.issuer.value == SAML_REPOSE_ISSUER
+
+        and: "the rest of the Assertions should have the original Issuer"
+        response.assertions.drop(1).every { it.issuer.value == SAML_EXTERNAL_ISSUER }
+
+        where:
+        sigOne | sigTwo | sigThree
+        true   | true   | true
+        false  | true   | true
+        true   | false  | true
+        true   | true   | false
+        false  | false  | false
+    }
+
+    def "a saml:response with an assertion missing the Issuer element should be rejected"() {
+        given: "a saml:response with an assertion missing the Issuer element"
+        def saml = samlResponse { MarkupBuilder builder ->
+            builder.'saml2:Issuer'(SAML_EXTERNAL_ISSUER)
+            builder.'saml2p:Status' {
+                'saml2p:StatusCode'(Value: SAML_STATUS_SUCCESS)
+            }
+            builder.'saml2:Assertion'(ID: "_" + UUID.randomUUID().toString(), IssueInstant: "2013-11-15T16:19:06.310Z", Version: "2.0") {
+                'saml2:Subject' {
+                    'saml2:NameID'(Format: "urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified", "john.doe")
+                    'saml2:SubjectConfirmation'(Method: "urn:oasis:names:tc:SAML:2.0:cm:bearer") {
+                        'saml2:SubjectConfirmationData'(NotOnOrAfter: "2113-11-17T16:19:06.298Z")
+                    }
+                }
+                'saml2:AuthnStatement'(AuthnInstant: "2113-11-15T16:19:04.055Z") {
+                    'saml2:AuthnContext' {
+                        'saml2:AuthnContextClassRef'("urn:oasis:names:tc:SAML:2.0:ac:classes:PasswordProtectedTransport")
+                    }
+                }
+            }
+            builder
+        }
+
+        when:
+        def mc = deproxy.makeRequest(
+                url: reposeEndpoint + SAML_AUTH_URL,
+                method: HTTP_POST,
+                headers: [(CONTENT_TYPE): CONTENT_TYPE_FORM_URLENCODED],
+                requestBody: asUrlEncodedForm((PARAM_SAML_RESPONSE): encodeBase64(saml)))
+
+        then: "the client gets back a bad response"
+        mc.receivedResponse.code as Integer == SC_BAD_REQUEST
```

Comment:
I added the check on the next line (which unfortunately won't show this comment as outdated now).